### PR TITLE
Generator should not care about import paths

### PR DESCRIPTION
**Pull request by abhinav**
Moved import path resolution logic upwards so that the Generator isn't munging paths to figure out import paths.

A `ServicePackage` method is provided which will be used during service code generation to import the service-specific sub packages.

Depends on https://github.com/thriftrw/thriftrw-go/pull/95

### CODE REVIEW

**Comment by abhinav** [comment]
CC @prashantv

File: gen/field.go [hunk 1]
```diff
@@ -28,6 +28,25 @@ type fieldGroupGenerator struct {
 	Fields compile.FieldGroup
 }
 
+func (f fieldGroupGenerator) Generate(g Generator) error {
```
Comment (prashantv) [inline gen/field.go]: let's move this commit into a separate PR since it seems unrelated?

File: gen/field.go [hunk 2]
```diff
@@ -28,6 +28,25 @@ type fieldGroupGenerator struct {
...
+	}
+
+	if err := f.FromWire(g); err != nil {
+		return err
+	}
+
+	if err := f.String(g); err != nil {
+		return err
+	}
+	return nil
```
Comment (prashantv) [inline gen/field.go]: nit: blank before this (to match the blank between the other groups in this method)

File: gen/generate.go [hunk 3]
```diff
@@ -71,41 +71,87 @@ func Generate(m *compile.Module, o *Options) error {
...
+
+// ServicePackage returns the import path for the package for the Thrift service
+// with the given name defined in the given file.
+func (i thriftPackageImporter) ServicePackage(file, name string) (string, error) {
+	topPackage, err := i.Package(file)
+	if err != nil {
+		return "", err
+	}
+
+	return filepath.Join(topPackage, strings.ToLower(name)), nil
```
Comment (prashantv) [inline gen/generate.go]: nit: should we do a "clean" of the name (which is lowercasing, removing any dashes or underscores etc?) or is just lowercase enough

File: gen/generate_test.go [hunk 4]
```diff
@@ -62,3 +62,51 @@ func TestGenerateWithRelativePaths(t *testing.T) {
...
+		},
+		{
+			File:     "/src/thrift/shared/common.thrift",
+			Relative: "shared/common",
+			Package:  "github.com/myteam/myservice/shared/common",
+		},
+	}
+
+	for _, tt := range tests {
+		if tt.Relative != "" {
```
Comment (prashantv) [inline gen/generate_test.go]: `Relative` seems to be always set?

File: gen/generate_test.go [hunk 5]
```diff
@@ -62,3 +62,51 @@ func TestGenerateWithRelativePaths(t *testing.T) {
...
+			got, err := importer.RelativePackage(tt.File)
+			if assert.NoError(t, err, "RelativePackage(%q)", tt.File) {
+				assert.Equal(t, tt.Relative, got, "RelativePackage(%q)", tt.File)
+			}
+		}
+
+		if tt.Package != "" {
```
Comment (prashantv) [inline gen/generate_test.go]: same with `Package`

File: gen/generate_test.go [hunk 6]
```diff
@@ -62,3 +62,51 @@ func TestGenerateWithRelativePaths(t *testing.T) {
...
+			got, err := importer.Package(tt.File)
+			if assert.NoError(t, err, "Package(%q)", tt.File) {
+				assert.Equal(t, tt.Package, got, "Package(%q)", tt.File)
+			}
+		}
+
+		if tt.ServicePackage != "" {
```
Comment (prashantv) [inline gen/generate_test.go]: nit: using `ServiceName` for this seems better since that's the input

**Comment by prashantv** [comment]
lgtm, just move the `Generate` fieldgroup changes out, and fix minor nits.

File: gen/generate_test.go (hunk 4 above)
Comment (abhinav) [inline gen/generate_test.go]: Yeah, for all existing test cases, but I'm treating it as optional. If Relative is provided, that's the expected `RelativePackage`, if `ServicePackage` is provided, that`s the expected `ServicePackage` and so on.

File: gen/generate_test.go (hunk 5 above)
Comment (abhinav) [inline gen/generate_test.go]: Answered above

File: gen/generate_test.go (hunk 6 above)
Comment (abhinav) [inline gen/generate_test.go]: `ServiceName` is the name passed to `ServicePackage`. This is the expected output of calling `ServicePackage(file, name)`.

File: gen/field.go (hunk 1 above)
Comment (abhinav) [inline gen/field.go]: Cool. I figured it was a small enough change to bundle in this PR.

File: gen/generate.go (hunk 3 above)
Comment (abhinav) [inline gen/generate.go]: Service names have to be valid identifiers so they won't contain spaces or dashes. Underscores may be present but I think that's okay to have because we want to support `service FooBar` and `service Foo_Bar` defined at the same time. We don't need to support `service foo` and `service Foo` defined at the same time because we validate case insensitivity at compile time.

File: gen/field.go (hunk 2 above)
Comment (abhinav) [inline gen/field.go]: :+1:

File: gen/generate.go (hunk 3 above)
Comment (prashantv) [inline gen/generate.go]: cool, makes sense. sgtm.
